File: golang-beer-game/application/OrderService.go
package application

import (
	"context"
	"errors"
	"fmt"
	"github.com/LeonFelipeCordero/golang-beer-game/application/events"
	"github.com/LeonFelipeCordero/golang-beer-game/domain"
	"github.com/LeonFelipeCordero/golang-beer-game/repositories/adapters"
	"github.com/google/uuid"
	"math/rand"
	"time"
)

type OrderService struct {
	repository    adapters.OrderRepositoryAdapter
	playerService PlayerService
	boardService  BoardService
	eventChan     chan events.Event
}

func NewOrderService(
	repository adapters.OrderRepositoryAdapter,
	playerService PlayerService,
	boardService BoardService,
	eventChan chan events.Event,
) OrderService {
	return OrderService{
		repository:    repository,
		playerService: playerService,
		boardService:  boardService,
		eventChan:     eventChan,
	}
}

func (o OrderService) CreateOrder(ctx context.Context, receiverId string) (*domain.Order, error) {
	receiver, err := o.playerService.Get(ctx, receiverId)
	if err != nil {
		return nil, err
	}
	sender, err := o.playerService.GetContraPart(ctx, *receiver)
	if err != nil {
		return nil, err
	}
	board, err := o.boardService.GetByPlayer(ctx, receiverId)
	if err != nil {
		return nil, err
	}

	order := domain.Order{
		Amount:         receiver.WeeklyOrder,
		OriginalAmount: receiver.WeeklyOrder,
		OrderType:      domain.OrderTypePlayerOrder,
		Status:         domain.StatusPending,
		Sender:         sender.Id,
		Receiver:       receiverId,
		CreatedAt:      time.Now().UTC(),
	}

	savedOrder, err := o.repository.Save(ctx, order)
	if err != nil {
		return nil, err
	}
	o.eventChan <- events.Event{
		Id:        uuid.NewString(),
		ObjectId:  board.Id,
		EventType: events.EventTypeNew,
		Object:    *savedOrder,
	}

	return savedOrder, nil
}

func (o OrderService) DeliverOrder(ctx context.Context, orderId string, amount int64) (*domain.Order, error) {
	order, err := o.repository.Get(ctx, orderId)

	var receiver *domain.Player
	if order.Receiver != "" {
		receiver, err = o.playerService.Get(ctx, order.Receiver)
		if err != nil {
			return nil, err
		}
	}

	sender, err := o.playerService.Get(ctx, order.Sender)
	if err != nil {
		return nil, err
	}

	board, err := o.boardService.GetByPlayer(ctx, sender.Id)
	if err != nil {
		return nil, err
	}

	// todo change this
	order.Amount = amount
	if !order.ValidOrderAmount() {
		return nil, errors.New("the new value can't be bigger than the original one")
	} else if !sender.HasStock(order.Amount) {
		return nil, errors.New("the Sender don't have enough stock to deliver this order")
	}

	savedOrder, err := o.repository.MarkAsFilled(ctx, orderId, amount)
	if err != nil {
		return nil, err
	}

	// todo last order?
	sender.Stock -= savedOrder.Amount
	if savedOrder.Receiver != "" {
		receiver.Stock += savedOrder.Amount
	}

	//todo do all saves in a single transaction
	if receiver != nil {
		o.playerService.Save(ctx, *receiver)

		o.eventChan <- events.Event{
			Id:        uuid.NewString(),
			ObjectId:  board.Id,
			EventType: events.EventTypeUpdate,
			Object:    *receiver,
		}
	}
	o.playerService.Save(ctx, *sender)
	//savedOrder, err := o.repository.Save(ctx, *order)
	if err != nil {
		return nil, err
	}

	o.eventChan <- events.Event{
		Id:        uuid.NewString(),
		ObjectId:  board.Id,
		EventType: events.EventTypeUpdate,
		Object:    *savedOrder,
	}

	o.eventChan <- events.Event{
		Id:        uuid.NewString(),
		ObjectId:  board.Id,
		EventType: events.EventTypeUpdate,
		Object:    *sender,
	}

	return savedOrder, nil
}

func (o OrderService) Get(ctx context.Context, orderId string) (*domain.Order, error) {
	return o.repository.Get(ctx, orderId)
}

func (o OrderService) LoadByBoard(ctx context.Context, boardId string) ([]domain.Order, error) {
	return o.repository.LoadByBoard(ctx, boardId)
}

func (o OrderService) LoadByPlayer(ctx context.Context, playerId string) ([]domain.Order, error) {
	return o.repository.LoadByPlayer(ctx, playerId)
}

func (o OrderService) DeliverFactoryBatch(ctx context.Context) {
	boards, err := o.boardService.GetActiveBoards(ctx)
	if err != nil {
		panic(fmt.Sprintf("error getting active boards %s", err))
	}
	for _, board := range boards {
		factory, _ := o.playerService.GetPlayersByRoleAndBoard(ctx, board.Id, string(domain.RoleFactory))
		if factory != nil {
			factory.Stock += factory.WeeklyOrder
			factory.Backlog += factory.WeeklyOrder
			factory.LastOrder = factory.WeeklyOrder
			o.playerService.Save(ctx, *factory)
			o.eventChan <- events.Event{
				Id:        uuid.NewString(),
				ObjectId:  board.Id,
				EventType: events.EventTypeUpdate,
				Object:    *factory,
			}
		}
	}
}

func (o OrderService) CreateCpuOrders(ctx context.Context) {
	boards, err := o.boardService.GetActiveBoards(ctx)
	if err != nil {
		panic(fmt.Sprintf("error getting active boards %s", err))
	}
	for _, board := range boards {
		o.createOrdersForBoard(ctx, board)
	}
}

func (o OrderService) createOrdersForBoard(ctx context.Context, board domain.Board) {
	players, err := o.playerService.GetPlayersByBoard(ctx, board.Id)
	if err != nil {
		panic(fmt.Sprintf("error getting players from board %s, %s", board.Id, err))
	}
	for _, player := range players {
		o.createOrdersForPlayer(ctx, player, board)
	}
}

func (o OrderService) createOrdersForPlayer(ctx context.Context, player domain.Player, board domain.Board) {
	num := rand.Intn(5-1) + 1

	for i := 0; i < num; i++ {
		order := domain.Order{
			Amount:         player.WeeklyOrder / 4,
			OriginalAmount: player.WeeklyOrder / 4,
			OrderType:      domain.OrderTypeCPUOrder,
			Status:         domain.StatusPending,
			Sender:         player.Id,
			CreatedAt:      time.Now().UTC(),
		}
		savedOrder, err2 := o.repository.Save(ctx, order)
		if err2 != nil {
			panic("Something went wrong creating cpu order")
		}
		o.eventChan <- events.Event{
			Id:        uuid.NewString(),
			ObjectId:  board.Id,
			EventType: events.EventTypeNew,
			Object:    *savedOrder,
		}
	}
}

File: beer-game-spring/src/main/kotlin/com/beer/game/application/order/OrderService.kt
package com.beer.game.application.order

import com.beer.game.application.board.BoardService
import com.beer.game.repositories.order.OrderStorageAdapter
import com.beer.game.common.OrderState
import com.beer.game.common.OrderType
import com.beer.game.common.Role
import com.beer.game.domain.*
import com.beer.game.domain.exceptions.ImpossibleActionException
import com.beer.game.application.events.InternalEventListener
import com.beer.game.application.player.PlayerService
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class OrderService(
    private val orderStorageAdapter: OrderStorageAdapter,
    private val boardService: BoardService,
    private val playerService: PlayerService,
    private val internalEventListener: InternalEventListener
) {

    fun createCpuOrders() {
        boardService.loadActiveBoards()
            .forEach { createCpuOrderForBoard(it) }
    }

    fun createOrder(receiverId: String): Pair<Order, String> {
        val board = boardService.loadBoardByPlayerId(receiverId)
        val receiver = board.findPlayer(receiverId)
        val sender = board.findContraPart(receiver)
        val order = Order(
            amount = receiver.weeklyOrder,
            originalAmount = receiver.weeklyOrder,
            type = OrderType.PLAYER_ORDER,
            sender = sender.id,
            receiver = receiver.id,
            createdAt = LocalDateTime.now()
        )
        sender.addOrder(order)
        receiver.addOrder(order)
        order.emitCreation(internalEventListener, board.id)
        return Pair(orderStorageAdapter.createOrder(board, order), board.id)
    }

    fun deliverOrder(orderId: String, amount: Int? = null) {
        val board = boardService.loadBoardByOrderId(orderId)
        val order = board.findOrder(orderId)
        amount?.let { order.amount = it }
        val sender = board.players.first { order.sender == it.id }
        val receiver = order.receiver?.let { board.players.first { order.receiver == it.id } }

        if (!order.validOderAmount()) {
            throw ImpossibleActionException(
                "deliver amount can't be bigger than original amount",
                "verify order amount"
            )
        }
        if (!sender.hasEnoughStock(order.amount)) {
            sender.emitUpdate(internalEventListener, board.id)
            throw ImpossibleActionException(
                "Sender doesn't have enough stock to deliver the order",
                "wait until the sender has enough stock, you can order more"
            )
        }

        // todo last order?
        sender.stock -= order.amount
        receiver?.let {
            it.stock += order.amount
            it.lastOrder = order.amount
        }
        order.state = OrderState.DELIVERED
        amount?.let { order.amount = amount }

        orderStorageAdapter.deliverOrder(order, board)
        order.emitUpdate(internalEventListener, board.id)
        sender.emitUpdate(internalEventListener, board.id)
        receiver?.apply { emitUpdate(internalEventListener, board.id) }
    }

    fun deliverFactoryBatches() {
        boardService.loadActiveBoards().forEach { deliverFactoryBatch(it) }
    }

    fun getOrder(orderId: String): Pair<Order, String> {
        val board = boardService.loadBoardByOrderId(orderId)
        val order = board.findOrder(orderId)
        return Pair(order, board.id)
    }

    fun getOrdersByBoard(boardId: String): List<Order> {
        return boardService.loadBoard(boardId)
            .players
            .flatMap { it.orders }
    }

    fun getOrdersByPlayer(playerId: String): List<Order> {
        val (player, _) = playerService.getPlayer(playerId)
        return player.orders
    }

    private fun createCpuOrderForBoard(board: Board) {
        board.players
            .forEach { player ->
                val order = Order(
                    amount = player.weeklyOrder / 4,
                    originalAmount = player.weeklyOrder / 4,
                    type = OrderType.CPU_ORDER,
                    sender = player.id,
                    createdAt = LocalDateTime.now()
                )
                player.addOrder(order)
                orderStorageAdapter.createOrder(board, order)
                order.emitCreation(internalEventListener, board.id)
            }
    }

    private fun deliverFactoryBatch(board: Board) {
        val factory = board
            .players
            .first { it.role == Role.FACTORY }
        factory.stock += factory.weeklyOrder
        factory.backlog += factory.weeklyOrder
        factory.lastOrder = factory.weeklyOrder
        orderStorageAdapter.deliverFactoryBatch(board)
        factory.emitUpdate(internalEventListener, board.id)
    }
}